hail/src/main/scala/is/hail/expr/ir/analyses/SemanticHash.scala

```diff
@@ -0,0 +1,433 @@
+           _: MatrixFilterRows |
+           _: MatrixMapCols |
+           _: MatrixFilterCols |
+           _: MatrixMapEntries |
+           _: MatrixFilterEntries |
+           _: MatrixDistinctByRow |
+           _: NDArrayShape |
+           _: NDArraySlice |
+           _: NDArrayReshape |
+           _: NDArrayWrite |
+           _: RelationalLet |
+           _: RNGSplit |
+           _: RNGStateLiteral |
+           _: StreamAgg |
+           _: StreamDrop |
+           _: StreamDropWhile |
+           _: StreamFilter |
+           _: StreamFlatMap |
+           _: StreamFold |
+           _: StreamFold2 |
+           _: StreamFor |
+           _: StreamIota |
+           _: StreamLen |
+           _: StreamMap |
+           _: StreamRange |
+           _: StreamTake |
+           _: StreamTakeWhile |
+           _: TableGetGlobals |
+           _: TableAggregate |
+           _: TableCollect |
+           _: TableCount |
+           _: TableDistinct |
+           _: TableFilter |
+           _: TableMapGlobals |
+           _: TableMapRows |
+           _: TableRename |
+           _: ToArray |
+           _: ToDict |
+           _: ToSet |
+           _: ToStream |
+           _: Trap =>
+        ()
+
+      // Discrete values
+      case _: Void | _: True | _: False =>
+        ()
+
+      // integral constants
+      case I32(x) => buffer ++= Bytes.fromInt(x)
+      case I64(x) => buffer ++= Bytes.fromLong(x)
+      case F32(x) => buffer ++= Bytes.fromFloat(x)
+      case F64(x) => buffer ++= Bytes.fromDouble(x)
+      case NA(typ) => buffer ++= EncodeTypename(typ)
+      case Str(x) => buffer ++= x.getBytes
+
+      // In these cases, just return None meaning that two
+      // invocations will never return the same thing.
+      case _ =>
+        return Left(s"SemanticHash unknown: ${ir.getClass.getName}")
+    }
+
+    Right(buffer.result())
+  }
+
+  def getFileHash(fs: FS)(path: String): Array[Byte] =
+    fs.eTag(path) match {
+      case Some(etag) =>
+        etag.getBytes
+      case None =>
+        path.getBytes ++ Bytes.fromLong(fs.fileStatus(path).getModificationTime)
+    }
+
+
+  def levelOrder(root: BaseIR): Iterator[(BaseIR, Int)] = {
+    val adj: ((BaseIR, Int)) => Iterator[(BaseIR, Int)] =
+      Function.tupled((ir, _) => ir.children.zipWithIndex.iterator)
+
+    TreeTraversal.levelOrder(adj)((root, 0))
+  }
+
+  object Bytes {
+    def fromInt(x: Int): Array[Byte] =
+      ByteBuffer.allocate(4).putInt(x).array()
+
+    def fromLong(x: Long): Array[Byte] =
+      ByteBuffer.allocate(8).putLong(x).array()
+
+    def fromFloat(x: Float): Array[Byte] =
+      ByteBuffer.allocate(4).putFloat(x).array()
+
+    def fromDouble(x: Double): Array[Byte] =
+      ByteBuffer.allocate(8).putDouble(x).array()
+
+    def fromClass(clz: Class[_]): Array[Byte] =
+      fromInt(clz.hashCode())
+  }
+
+  object CodeGenSupport {
+    def lift(hash: SemanticHash.Type): Option[SemanticHash.Type] =
+      Some(hash)
+  }
```

Comment:
I was having a horrible time trying to generate code to lift an `Int` into an `Option` and kept getting `NoSuchMethodError`s when trying to call the constructor of `Some` or call `Option.apply`, presumably because `Option` is parameterised and thus needs a reference type in its jvm implementation.
I rage quit and wrote this instead.